### PR TITLE
cluster manager: add optional host predicate to drainConnections()

### PR DESCRIPTION
**Pull request by mattklein123**
This will allow EM to selectively drain connections based on hostname(),
for example.

Risk Level: Low
Testing: New integration test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: test/integration/http_conn_pool_integration_test.cc [hunk 1]
```diff
@@ -106,14 +106,43 @@ TEST_P(HttpConnPoolIntegrationTest, PoolDrainAfterDrainApiSpecificCluster) {
...
+  auto response = codec_client_->makeRequestWithBody(default_request_headers_, 1024);
+  waitForNextUpstreamRequest();
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+  upstream_request_->encodeData(512, true);
+  ASSERT_TRUE(response->waitForEndStream());
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_TRUE(response->complete());
+
+  // Perform a drain request which doesn't actually do a drain.
+  test_server_->server().dispatcher().post([this] {
```
Comment (alyssawilk) [inline test/integration/http_conn_pool_integration_test.cc:131]: probably worth doing a second drain which matches.  bonus points if we have a unit test somewhere for matches A and not B.

**Review by alyssawilk — APPROVED** [review]
code LGTM but I'd like a bit more testing, either here or in a follow up.

**Review by alyssawilk — APPROVED** [review]
thanks!

**Comment by mattklein123** [comment]
I'm still trying to figure out a unit test but it's not so easy. I might merge this and then do another PR but I will see how it goes.

**Review by RyanTheOptimist — APPROVED** [review]

**Pull request merged**